### ISSUE TITLE
Go: Free functions to structures to decoupled dependencies

### ISSUE THREAD

**Issue by AmitKumarDas** [AmitKumarDas/Decisions#197]
Let us start of with the following code snippet that I have taken from project [metac](https://github.com/AmitKumarDas/metac)

```go
// ManageChildren ensures the relevant children objects of the
// given parent are in sync
func ManageChildren(
	dynClient *dynamicclientset.Clientset,
	updateStrategy ChildUpdateStrategy,
	parent *unstructured.Unstructured,
	observedChildren, desiredChildren ChildMap,
) error {
	// If some operations fail, keep trying others so, for example,
	// we don't block recovery (create new Pod) on a failed delete.
	var errs []error

	// Delete observed, owned objects that are not desired.
	for key, objects := range observedChildren {
		apiVersion, kind := ParseChildMapKey(key)
		client, err := dynClient.Kind(apiVersion, kind)
		if err != nil {
			errs = append(errs, err)
			continue
		}
		err := deleteChildren(client, parent, objects, desiredChildren[key])
                if err != nil {
			errs = append(errs, err)
			continue
		}
	}

	// Create or update desired objects.
	for key, objects := range desiredChildren {
		apiVersion, kind := ParseChildMapKey(key)
		client, err := dynClient.Kind(apiVersion, kind)
		if err != nil {
			errs = append(errs, err)
			continue
		}
		if err := updateChildren(
			client,
			updateStrategy,
			parent,
			observedChildren[key],
			objects,
		); err != nil {
			errs = append(errs, err)
			continue
		}
	}

	return utilerrors.NewAggregate(errs)
}
```

So what is above snippet trying to do? We shall try to form some understanding without getting
into the details. These are my observations:
- It is a function that takes in a lot of arguments
  - We are calling this as a free function i.e. it is not tied to any structure
- It loops across one of the passed arguments & invokes delete call for each iteration
- It loops across another passed argument & invokes update call for each iteration
- It does not stop in case of errors but aggregates them & returns it as an error instance

The next set of things that occurs to me & hopefully to you as well are the following aspects of programming:
- Is the code readable?
- Is the code maintainable?
- Is the code unit testable?

**Comment by AmitKumarDas** [comment]
Let us pick up from the earlier comment that had a set of questions. In my opinion, the code is readable to a Go programmer. With regards to maintainability, if this code will never change in due course of releases, we can assume nil maintenance. However, this hypothesis can be proved only on hindsight i.e. in the future releases. Given a choice I am typically paranoid about maintenance and will go out of my way to factor in all the maintainability aspects. In other words, I will ensure my code adheres to `Open Closed Principle`. Lets talk a bit about unit testing aspects. I believe this is not a great example of code that can be unit tested. For example, invocations to `deleteChildren` & `updateChildren` will hinder the unit testing aspects of `ManageChildren`.

To summarise I will vote the questions with below:
- Is the code readable? :+1: 
- Is the code maintainable? :-1: 
- Is the code unit testable? :no_entry_sign:

**Comment by AmitKumarDas** [comment]
Now that we have gone through the code snippet and also have talked a bit about various programming aspects on this code, lets try to figure out the solutions that can pass all the above checks.

In my opinion, if a function tries to accept more than two instances (it might be interfaces or concrete structures or combinations), we should start thinking of a structure and convert this function to a method of this structure. This will help us to follow the `Open Closed Principle` that we talked about in earlier comment.

```go
// BaseManager holds the properties required for various controller
// manager operations
type BaseManager struct {
	// UpdateStrategy to be followed while executing update
	// operation against a child resource
	UpdateStrategy ChildUpdateStrategy

	// Resource that is being watched
	Watched *unstructured.Unstructured

	// IsWatchOwner indicates if the watched resource is also the
	// owner (of the child resources declared in the controller)
	IsWatchOwner bool
}

// ControllerManager manages applying metac controller child resources 
// in Kubernetes cluster. Here apply implies either Create or Update or 
// Delete of child resources against a given Kubernetes cluster.
type ControllerManager struct {
	BaseManager

	// DynClientSet is responsible to provide a dynamic client for
	// a specific resource
	DynClientSet *dynamicclientset.Clientset

	// Observed state (i.e. current state in kubernetes) of child 
	// resources referred to in the controller
	Observed ChildMap

	// Desired kubernetes state of child resources referred to in
	// the controller
	Desired ChildMap
}
```
```go
func (m *ControllerManager) MakeDeleteManagers() (ChildrenListDeleter, error) {
	var errs []error
	var mgrs []*ChildrenManager

	// Delete observed, owned objects that are not desired.
	for verkind, objects := range m.Observed {
		apiVersion, kind := ParseChildMapKey(verkind)
		client, err := m.DynClientSet.Kind(apiVersion, kind)
		if err != nil {
			errs = append(errs, err)
			continue
		}
		cd := &ChildrenDeleter{
			BaseManager: m.BaseManager,
			DynClient: client,
			Observed: objects,
			Desired: m.Desired[verkind],
		}
		mgrs = append(mgrs, cd)
	}

	return ChildrenListDeleter(mgrs), utilerrors.NewAggregate(errs)
}

func (m *ControllerManager) MakeUpdateManagers() (ChildrenListUpdater, error) {
	var errs []error
	var mgrs []*ChildrenManager

	// Create or update desired objects
	for verkind, objects := range m.Desired {
		apiVersion, kind := ParseChildMapKey(verkind)
		client, err := m.DynClientSet.Kind(apiVersion, kind)
		if err != nil {
			errs = append(errs, err)
			continue
		}
		cu := &ChildrenUpdater{
			BaseManager: m.BaseManager,
			DynClient: client,
			Observed: m.Observed[verkind],
			Desired: objects,
		}
		mgs = append(mgrs, cu)
	}

	return ChildrenListUpdater(mgrs), utilerrors.NewAggregate(errs)
}

// Apply executes create, delete or update operations against
// the child resources set against this manager instance
func (m *ControllerManager) Apply() error {
	var errs []error

	dmlist, err := m.MakeDeleteManagers()
	if err != nil {
		errs = append(errs, err)
	}
	err = dmlist.Delete()
	if err != nil {
		errs = append(errs, err)
	}

	umlist, err := m.MakeUpdateManagers()
	if err != nil {
		errs = append(errs, err)
	}
	err = umlist.Update()
	if err != nil {
		errs = append(errs, err)
	}

	return utilerrors.NewAggregate(errs)
}
```
```go
type ChildrenManager struct {
	BaseManager

	DynClient *dynamicclientset.ResourceClient
	Observed map[string]*unstructured.Unstructured
	Desired map[string]*unstructured.Unstructured
}

type ChildrenListDeleter []*ChildrenManager
type ChildrenListUpdater []*ChildrenManager

func (d ChildrenListDeleter) Delete() error {
  //..
}

func (d ChildrenListUpdater) Update() error {
  //..
}
```

**Comment by AmitKumarDas** [comment]
One may argue that above change has only tried to break original logic into smaller pieces i.e. modular code. Does modular code answer all the queries we talked about?

The answer is `No` if we just think in terms of modular code. We need to realise that above change is about moving free functions to their structural implementations. I will talk about other changes  in the next set of comments that explains the power that this change brings in. 

Meanwhile, let us try to summarise on the code changes so far:
- We created structures. In fact a lot of them since the requirement demands it.
- We favoured composition/embedding of structures to arrive at new structures
- Each structure has its own role to play i.e. each will get its own method receiver.
- Above changes did not impact the readability aspects
- Moving to a structure helps us to add new properties to this structure if required
  - This avoids changing the function signature now
  - Which in turn implies no change to existing callers

NOTES:
- On the whole we tried to create/design a structure for every function that accepted a lot of instances. - We modularised the original `ManageChildren` into:
  - `Apply`, 
  - `MakeDeleteManagers` & 
  - `MakeUpdateManagers`. 
- Modular code enabled decoupling of building the updaters & deleters from actual update & delete invocations.

Let us try to vote this refactored code:
- Is the code readable? :+1: 
- Is the code maintainable? :+1: 
  - This is definitely more maintainable than earlier one
  - This is more aligned to Open Closed Principle
- Is the code unit testable? :confused: 
  - How do we mock delete & update invocations within Apply in-order to unit test Apply?

**Comment by AmitKumarDas** [comment]
Let us now see if we can unit test Apply as per above refactored code

```go
// in your `_test.go` file

// FakeControllerManagerForApply is a type defined out of ControllerManager
type FakeControllerManagerForApply ControllerManager

// override real implementation
func (mock *FakeControllerManagerForApply) MakeDeleteManagers()
 (ChildrenListDeleter, error) {
  return ChildrenListDeleter{}, nil
}

// override real implementation
func (mock *FakeControllerManagerForApply) MakeUpdateManagers() 
 (ChildrenListUpdater, error) {
  return ChildrenListUpdater{}, nil
}

func TestApply(t *testing.T) {
  // make use of FakeControllerManagerForApply to test the logic present in 
  // ControllerManager's Apply
}
```

Here, we have tried to define more types against our own structures. In my opinion, this is a simple approach versus one that defines interfaces and third party mocking libraries to solve the unit test problem.

Thus, with this approach of thinking in terms of structures & types we are able to achieve all our problem statements.
